File: src/Services/Product/Product.API/Endpoints/ListProductCategories.cs
using BuildingBlock.Application.Abstractions.Common;
using BuildingBlock.Infrastructure.Authorization;
using BuildingBlock.SharedKernel.Extensions;

using Product.Application.Features.ProductCategories.Queries.ListProductCategories;

namespace Product.API.Endpoints;

public sealed class ListProductCategoriesEndpoint : ICarterModule
{
    private readonly string[] API_DESC = [
        "## List Product Categories",
        "",
        "Returns the full flat category set (Id/ParentCategoryId pairs) so a client can assemble",
        "the hierarchy tree itself. Not paginated - category counts are small reference data.",
    ];

    public void AddRoutes(IEndpointRouteBuilder app)
    {
        app.MapGet("/categories", Handle)
            .RequireAuthorization(AuthorizationPolicies.RequireAuthenticated)
            .WithName("ListProductCategories")
            .WithDisplayName("List Product Categories API")
            .WithDescription(API_DESC.JoinToString("\n"))
            .Produces<ApiResponse<ListProductCategoriesResponse>>(StatusCodes.Status200OK);
    }

    private static async Task<IResult> Handle(
        [FromServices] ISender sender,
        CancellationToken ct = default)
    {
        var response = await sender.Send(new ListProductCategoriesQuery(), ct);

        return Results.Ok(ApiResponse<ListProductCategoriesResponse>.Ok(response));
    }
}

File: src/services/product/list-product-categories.ts
import { apiClient } from '@/shared/lib/api/client';
import { BASE_PATH } from './_base';
import type { ProductCategoryStatus } from './types/product-category-status';

export interface ProductCategoryItemResponse {
  id: string;
  code: string | null;
  name: string | null;
  status: ProductCategoryStatus;
  parentCategoryId: string | null;
}

export interface ListProductCategoriesResponse {
  categories: ProductCategoryItemResponse[] | null;
}

/** GET /categories — full flat set, not paginated (small reference data). */
export function listProductCategories(): Promise<ListProductCategoriesResponse> {
  return apiClient.get(`${BASE_PATH}/categories`).then((res) => res.data);
}
